substrate/client/network/sync/src/lib.rs

```diff
@@ -2984,54 +2980,51 @@ fn peer_gap_block_request<B: BlockT>(
 /// Get pending fork sync targets for a peer.
 fn fork_sync_request<B: BlockT>(
 	id: &PeerId,
-	targets: &mut HashMap<B::Hash, ForkTarget<B>>,
+	targets: &mut BTreeMap<(NumberFor<B>, B::Hash), ForkTarget<B>>,
 	best_num: NumberFor<B>,
 	finalized: NumberFor<B>,
 	attributes: BlockAttributes,
 	check_block: impl Fn(&B::Hash) -> BlockStatus,
 	max_blocks_per_request: u32,
 ) -> Option<(B::Hash, BlockRequest<B>)> {
-	targets.retain(|hash, r| {
-		if r.number <= finalized {
-			trace!(target: "sync", "Removed expired fork sync request {:?} (#{})", hash, r.number);
+	targets.retain(|(number, hash), _| {
+		if *number <= finalized {
+			trace!(target: "sync", "Removed expired fork sync request {:?} (#{})", hash, number);
 			return false
 		}
 		if check_block(hash) != BlockStatus::Unknown {
-			trace!(target: "sync", "Removed obsolete fork sync request {:?} (#{})", hash, r.number);
+			trace!(target: "sync", "Removed obsolete fork sync request {:?} (#{})", hash, number);
 			return false
 		}
 		true
 	});
-	for (hash, r) in targets {
+	// Download the fork only if it is behind or not too far ahead our tip of the chain
+	// Otherwise it should be downloaded in full sync mode.
+	let range = ..(best_num.saturating_add(max_blocks_per_request.into()), Default::default());
+	// Iterate in reverse order to download target with the highest number first.
+	// Other targets might belong to the same fork, so we don't need to download them separately.
+	for ((number, hash), r) in targets.range(range).rev() {
```

Comment:
Could you explain this logic in more detail? It checks the fork targets in reverse order by `(number, hash)` so the last block of the fork is checked first and if it's not available, it proceeds to the parent of the latest block, am I understanding this correctly?

Reply:
Not quite. The logic goes like this: 

If there are multiple `fork_targets`, in what order should they be checked? 
I believe the optimal approach would be to begin with the `fork_target` with the highest number (aka latest). If it's available, then download (number - finalized) blocks. If not, go to the next `fork_target` (can be unrelated with the first one).
The optimization lies in the fact that other `fork_targets` with lower numbers might be part of the same fork that has already been downloaded.

Reply:
I haven't tested running the original test yet with this optimization applied but how much better it got, do you remember?

It looks like it won't even traverse through the fork targets that are too far in to the future which is what we want but I think this could be optimized even further. Right now if a peer is not able to provide any fork targets, it still has to go through all of them which is not optimal. Which fork targets the peer is able to provide should probably be stored so the list of targets wouldn't have to be looped through each time we call `fork_sync_request()`. This will admittedly complicate the design and as @dmitry-markin is already working on the syncing code, it may be preferable to not make significant refactors until we've decided how the `ChainSync` should be structured.

Reply:
> how much better it got, do you remember?

In my tests, the difference was negligible, as the time spent in `fork_sync_request` is in microseconds even for 10k blocks.